librz/core/linux_heap_glibc.c

```diff
@@ -126,25 +136,258 @@ static GHT GH(get_main_arena_with_symbol)(RzCore *core, RzDebugMap *map) {
+	}
+
+cleanup:
+
+	rz_bin_file_delete(bin, libc_buf);
+	rz_bin_file_set_cur_binfile(bin, current_bf);
+	return buf;
+}
+
+RZ_API double GH(rz_get_glibc_version)(RzCore *core, const char *libc_path,
+									   const ut8 *banner_start) {
+	double version = 0.0;
+	
+	if (!banner_start) {
+		banner_start = GH(get_glibc_banner)(core, ".rodata", libc_path);
+		if (!banner_start) {
+			goto error;
+		}
+	}
+
+	const char *pattern = "release version (\\d.\\d\\d)";
+	RzRegex *re = rz_regex_new(pattern, RZ_REGEX_EXTENDED | RZ_REGEX_CASELESS, 0);
+	RzPVector *matches = rz_regex_match_first(re, (const char *)banner_start,
+											  RZ_REGEX_ZERO_TERMINATED, 0 , RZ_REGEX_DEFAULT);
+	RzRegexMatch *match = rz_pvector_at(matches, 1);
+	char *version_str = rz_str_ndup((const char*)banner_start + match->start, match->len);
+	if (!version_str) {
+		rz_pvector_free(matches);
+		return version;
+	}
+	version = strtod(version_str, NULL);
+	rz_pvector_free(matches);
+	if (version != 0) {
+		RZ_LOG_INFO ("libc version %.2f identified from .rodata banner", version);
+		return version;
+	}
+
+error:
+	RZ_LOG_WARN ("get_glibc_version failed");
+	return version;
+}
+
+static RzList *GH(fetch_tls_address_list)(RzDebug *dbg) {
+	RzList *t_list = NULL;
+
+	if (!dbg->threads) {
+		return t_list;
+	}
+
+	RzDebugPid *th;
+	RzListIter *it;
+
+	rz_list_foreach (dbg->threads, it, th) {
+		if (dbg->pid == th->pid) {
+			if (!th->tls) {
+				t_list = get_pid_thread_list(dbg, dbg->pid);
+				break;
+			}
+		}
+	}
+
+	return t_list;
+}
+
+static GHT GH(read_val)(RzCore *core, const void *src, bool is_big_endian) {
+	if (SZ == RZ_SYS_BITS_16) {
+		return rz_read_ble16(src, is_big_endian);
+	} else if (SZ == RZ_SYS_BITS_32) {
+		return rz_read_ble32(src, is_big_endian);
+	} else {
+		return rz_read_ble64(src, is_big_endian);
+	}	
+}
+
+static RZ_BORROW RzList *GH(fill_tcache_entries)(RzCore *core, GH(RTcache) * tcache) {
+	RzList *tcache_bins_list = rz_list_newf((RzListFree)GH(rz_heap_bin_free));
+
+	// Use rz_tcache struct to get bins
+	for (int i = 0; i < TCACHE_MAX_BINS; i++) {
+		int count = GH(tcache_get_count)(tcache, i);
+		GHT entry = GH(tcache_get_entry)(tcache, i);
+
+		RzHeapBin *bin = RZ_NEW0(RzHeapBin);
+		if (!bin) {
+			goto error;
+		}
+		bin->type = rz_str_dup("Tcache");
+		bin->bin_num = i;
+		bin->chunks = rz_list_newf((RzListFree)GH(rz_heap_chunk_free));
+		rz_list_append(tcache_bins_list, bin);
+		if (count <= 0) {
+			continue;
+		}
+		bin->fd = (ut64)(entry - GH(HDR_SZ));
+		// get first chunk
+		RzHeapChunkListItem *chunk = RZ_NEW0(RzHeapChunkListItem);
+		if (!chunk) {
+			goto error;
+		}
+		chunk->addr = (ut64)(entry - GH(HDR_SZ));
+		rz_list_append(bin->chunks, chunk);
+
+		if (count <= 1) {
+			continue;
+		}
+
+		// get rest of the chunks
+		GHT tcache_fd = entry;
+		GHT tcache_tmp = GHT_MAX;
+		for (size_t n = 1; n < count; n++) {
+			int r = rz_io_nread_at(core->io, tcache_fd, (ut8 *)&tcache_tmp, sizeof(GHT));
```

Comment:
this is not endian aware.
You should probably implement this by reading into a `ut8` buffer and then use the `rz_endian.h` functions.

Reply:
Updated with `read_le`